pallets/crowdloans/src/lib.rs

```diff
@@ -645,26 +674,33 @@ pub mod pallet {
             let mut migrated_count = 0u32;
             let mut all_migrated = true;
 
-            for (who, (amount, _)) in contributions {
+            // single migration has a processing limit
+            for (who, (amount, referral_code)) in contributions {
                 if migrated_count >= T::MigrateKeysLimit::get() {
                     all_migrated = false;
                     break;
+                } else {
+                    Self::do_migrate_contribution(
+                        &who,
+                        &mut vault,
+                        amount,
+                        ChildStorageKind::Pending,
+                        ChildStorageKind::Flying,
+                    )?;
+                    Self::do_contribute(&who, crowdloan, amount, referral_code)?;
+                    migrated_count += 1;
                 }
-                Self::do_migrate_contribution(
-                    &who,
-                    &mut vault,
-                    amount,
-                    ChildStorageKind::Pending,
-                    ChildStorageKind::Flying,
-                )?;
-                Self::do_contribute(&who, crowdloan, amount)?;
-                migrated_count += 1;
             }
 
             if all_migrated {
                 Self::deposit_event(Event::<T>::AllMigrated(crowdloan));
             } else {
-                Self::deposit_event(Event::<T>::PartiallyMigrated(crowdloan));
+                Self::deposit_event(Event::<T>::PartiallyMigrated(
+                    crowdloan,
+                    Self::contribution_iterator(vault.trie_index, ChildStorageKind::Pending)
```

Comment:
this will do 2nd full read of `pending contributions`, we should read only once

Reply:
but it doesn't implement copy trait

Reply:
but it doesn't implement Copy trait

Reply:
@rjman-ljm I mean `contribution_iterator` has been called twice

Reply:
I need to do `let total_count = contributions.count().saterated_into::<u32>();`, `count()` will take ownership and consume the iterator, so I have to call again to get a iterator.